### PR TITLE
Save detection samples

### PR DESCRIPTION
**Pull request by ivelin**
This PR enables saving detection samples for offline analysis and custom model training.

Example optional config parameters:

```
    ################
    # Regions specify where to look for detections in the camera images.
    # size is the size of a the region square.
    # x_offset and y_offset are regiona start corner relative to the 0,0 coordinates of the camera image.
    # min_person_area sets the minimum area size (width*height) for a detection to be interesting.
    # threshold sets the minimum confidence of the AI that the detected object is of the suggested category (label).
    # save_samples is an optional parameter that enable saving of regions with detected objects to local files.
    # The name format of saved files is :
    #  /data/<camera_name>/detections/timestamp.jpg for the input image (region of the camera image)
    #  /data/<camera_name>/detections/timestamp.json for the detections parameters
    # detections will be spaced at detections_interval seconds or longer to avoid near duplicate images.
    # Default is 2 seconds.
    # non_detections_interval indicates how often to save images where nothing was detected.
    # These can be used later for offline analysis and AI model training.
    # Default non_detections_interval is 300 seconds (5 minutes).
    ################
    regions:
      - size: 700
        x_offset: 250
        y_offset: 10
        min_person_area: 1000
        threshold: 0.6
        save_samples:
          detections_interval: 1
          non_detections_interval: 300

```

Added a related data_dir parameter to the tensorflow section:

```
################
# Optional Tensorflow settings. Change these only if you are familiar with Tensorflow.
# Intended for developer experimentation.
################
tensorflow:
  # object detection model and label paths
  # model: /config/mobilenet_ssd_v2_coco_quant_postprocess_edgetpu.tflite
  # labels: /config/coco_labels.txt
  # face detection model without labels
  model: /config/mobilenet_ssd_v2_face_quant_postprocess_edgetpu.tflite
  labels: /config/coco_labels.txt
  # Directory where output data will be persisted as needed. Default is /data/
  # For example images with detected object boxes can be saved in this directory. See also regions: section.
  # data_dir: /data/

```

When this feature is enabled in the config file, the data directory should be passed to the docker image:

via command line:
`-v /opt/frigate/data:/data:rw \`

or docker-compose.yml
```
    volumes:
      - /opt/frigate/data:/data
    ...
```

The resulting directory structure with saved samples looks like the screenshot below. Notice that the saved images correspond to detection regions, cropped from the original full camera capture. In other words these are the images that Tensorflow sees and can be used for custom model training with transfer learning:

Screen Shot 2019-08-24 at 4 31 09 PM

The JSON file with detection arguments has the same file prefix as the image file but .json extension. It  enumerates all detections in the given region snapshot. An example corresponding to the image above  looks like this:

```
[
    {
        "time": "2019-08-24 21:16:06.040360",
        "label_id": 0,
        "name": "person",
        "score": 0.69140625,
        "xmin": 3,
        "ymin": 137,
        "xmax": 207,
        "ymax": 389
    }
]
```

### CODE REVIEW

File: frigate/object_detection.py [hunk 1]
```diff
@@ -22,46 +29,150 @@ def ReadLabelFile(file_path):
...
 
         threading.Thread.__init__(self)
         self.cameras = cameras
         self.prepped_frame_queue = prepped_frame_queue
         
+        # print("tf_args: " + str(tf_args))
         # Load the edgetpu engine and labels
-        self.engine = DetectionEngine(PATH_TO_CKPT)
-        self.labels = ReadLabelFile(PATH_TO_LABELS)
+        if tf_args != None:
```
Comment (blakeblackshear) [inline frigate/object_detection.py]: Passing in different label files and models is already supported. You would do that with a Docker volume by passing `-v path/to/model/on/host.tflite:/frozen_inference_graph.pb` to the docker command or by specifying in your compose file for both the model and the labels.

File: frigate/object_detection.py [hunk 2]
```diff
@@ -22,46 +29,150 @@ def ReadLabelFile(file_path):
...
-                            'score': float(obj.score),
-                            'xmin': int((box[0] * frame['region_size']) + frame['region_x_offset']),
-                            'ymin': int((box[1] * frame['region_size']) + frame['region_y_offset']),
-                            'xmax': int((box[2] * frame['region_size']) + frame['region_x_offset']),
-                            'ymax': int((box[3] * frame['region_size']) + frame['region_y_offset'])
-                        })
-            self.cameras[frame['camera_name']].add_objects(parsed_objects)
+            try:
+                # print("Calling TF Coral to detect objects")
+                detected_objects = self.engine.DetectWithImage(
```
Comment (blakeblackshear) [inline frigate/object_detection.py]: Using `DetectWithImage` means that this thread is spending time resizing, cropping, and then converting this to a numpy array prior to detecting objects. All the time spent doing that is time that the Coral is not detecting objects, and will reduce the total possible speed of the Coral.

File: frigate/object_detection.py [hunk 3]
```diff
@@ -22,46 +29,150 @@ def ReadLabelFile(file_path):
...
+                    top_k=3)
+                #print("TF Coral detected {} objects withing {} ms".
+                #      format(len(objects),
+                #             self.engine.get_inference_time()))
+
+                # parse and pass detected objects back to the outbound camera stream
+                self.__pass_detections_to_output_cam(frame, detected_objects)
+
+                # save detections to file if enabled
+                self.__save_sample(frame, detected_objects)
```
Comment (blakeblackshear) [inline frigate/object_detection.py]: Because this is done on the same thread as the Coral, it must finish saving the files to disk before it can begin working on the next image. That will also reduce the FPS of the Coral.

File: frigate/object_detection.py [hunk 4]
```diff
@@ -90,23 +202,26 @@ def run(self):
...
-            # Expand dimensions since the model expects images to have shape: [1, 300, 300, 3]
-            frame_expanded = np.expand_dims(cropped_frame, axis=0)
 
+#            # Resize to 300x300 if needed
+#            if cropped_frame.shape != (300, 300, 3):
+#                cropped_frame = cv2.resize(cropped_frame, dsize=(300, 300), interpolation=cv2.INTER_LINEAR)
+#           # Expand dimensions since the model expects images to have shape: [1, 300, 300, 3]
+#           frame_expanded = np.expand_dims(cropped_frame, axis=0)
+
+            img = Image.fromarray(cropped_frame)
```
Comment (blakeblackshear) [inline frigate/object_detection.py]: What is happening here is the numpy array is converted to a PIL image, and then tensorflow converts it back to a 300x300 numpy array before detection. All that just adds processing delays.

**Review by blakeblackshear — COMMENTED** [review]
I really appreciate the PR, but I spent a lot of time measuring and optimizing for the maximum processing speed of the Coral. I like the way you are saving detections, but it really needs to be done on a separate thread so it doesn't block processing the next image. I am trying to do as little as possible inside of the PreppedQueueProcessor's while loop so the Coral can get to the next image as fast as possible.

File: frigate/object_detection.py (hunk 1 above)
Comment (ivelin) [reply frigate/object_detection.py]: Yes, that is true. Passing a custom graph via docker parameter works fine. 

Let me elaborate why I suggest adding config parameters as an option:

1. It allows to add supplementary parameters that might be useful. At the moment I only added /data dir for saved samples, but I anticipate potentially inference hyperparameters as we play with different models.

2. In the next iteration I was hoping we can allow multiple models to be used in one frigate run. I saw folks asking on the forums how they can run multiple models. Different models may be applicable for different regions or cameras. My tests with Coral show that switching between models has a minimal impact of about 10ms. It seems reasonable in the bigger picture. Apparently if multiple Corals are plugged-in, the TF Lite framework knows how to load balance, cache and optimize model reuse. I haven't tried that yet myself.

Hope that makes sense. Happy to jump on a call to brainstorm a bit. Ping me when time allows: ivelin117@gmail.clom

**Comment by ivelin** [comment]
> I really appreciate the PR, but I spent a lot of time measuring and optimizing for the maximum processing speed of the Coral. I like the way you are saving detections, but it really needs to be done on a separate thread so it doesn't block processing the next image. I am trying to do as little as possible inside of the PreppedQueueProcessor's while loop so the Coral can get to the next image as fast as possible.

Thanks for the review. It makes sense. I picked the faster route here because the performance impact seemed minimal on my RPI. Didn't want to overengineer it in the first go. Since you are OK with the feature, I will look at pulling the logic in a separate thread next.

File: frigate/object_detection.py (hunk 2 above)
Comment (ivelin) [reply frigate/object_detection.py]: Yes, that is correct. I was weighing convenience and performance here. The face model has a different input tensor shape (320x320 vs 300x300 for coco) which requires a bit more code to detect and resize. I am ok with writing the extra code, although I'm not sure if it will add significant performance benefit.

While we are on the topic of resizing. What are your thoughts on relaxing the square requirement for regions? My tests show that passing a rectangular region with preserved aspect ratio works well.

Comment (blakeblackshear) [reply frigate/object_detection.py]: It is a significant performance benefit when you are trying to run detection on more than 5 cameras with lots of regions and anything above 1fps. I have tried it both ways. Because the model is trained on squares, the only way you can maintain the aspect ratio is to crop the image. We can relax the requirement, but I would want to crop in the separate thread where resizing happens so this thread doesn't have to calculate it.

File: frigate/object_detection.py (hunk 1 above)
Comment (blakeblackshear) [reply frigate/object_detection.py]: When the time comes to support multiple models, then I would want to add the ability to pass them as parameters. It would likely be implemented in a different way for that use case anyway. For now, I don't think it is necessary.
